### ISSUE TITLE
Provide a `get-pip.py`-like script for virtualenv

### ISSUE THREAD

**Issue by asottile** [pypa/virtualenv#1092]
I often find myself on a machine with near-zero permissions (or a chroot / docker container) and have used this script: https://github.com/asottile/asottile.github.io#bootstrap-a-virtualenv-from-nothing

Would love to get this or something like this hosted similar to `get-pip.py`

**Comment by asottile** [comment]
I still think this is a good idea

**Comment by gaborbernat** [comment]
@asottile make a PR with tests, add it to the documentation and we're ready then 👍

**Comment by asottile** [comment]
sounds good! as for design there's kind of two options -- which would you prefer:

1. like `get-pip.py`: bundle the actual bits in the file itself
2. like my script I linked above, reach out to pypi to download the tgz / wheel



(1)  seems like a better idea but would like to get your opinion before working on it

**Comment by gaborbernat** [comment]
I would prefer the second, cleaner. Note we might need to call a wheel build though, going ahead we no longer support sdist evaluation (as I'd like the horrible one file we have now split up and concatenated during build rather than on the repo, this would drop need of doing current embedded hack). That being said would it be possible to use pypi wheel as download as opposed to a GitHub file?

**Comment by asottile** [comment]
option 1 is nice though if you don't have a network connection (think: vendor file in repo, able to create venvs without dependencies) -- doing the wheel would be fine as well (I was just using the sdist in my script for no good reason at the time)

**Comment by gaborbernat** [comment]
I would still go down the path of wheel extract, python can natively unzip it. I mean the user will have to get the repo in some way or another, so some network must exist, not?

**Comment by asottile** [comment]
Unless you're on macos :( it's crippled ssl implementation can't talk to pypi by default with just stdlib

**Comment by gaborbernat** [comment]
isn't github also https?

**Comment by asottile** [comment]
Yes but (for now) doesn't require tls 1.2 or whatever

**Comment by asottile** [comment]
And `curl` can take advantage of the other ssl implementation available on macos (pip does this) but urllib hasn't figured that out yet

**Comment by gaborbernat** [comment]
Oh I see, in that case we just need to use either pip or curl, not?

**Comment by asottile** [comment]
Then it fails the "bootstrap from just python" aspect :S

**Comment by asottile** [comment]
yeah, but once you have the bootstrapper it should be able to do everything itself -- even my script above fails that category until you've got a decently-compiled python on macos (brew for instance)

this is why I'm leaning towards option 1

**Comment by gaborbernat** [comment]
Isn't the wheel (perhaps wrapped beside a simple python script) able to do everything?

**Comment by gaborbernat** [comment]
I'm on board with that. It could be another artefact we build upon release? Should we publish it on PyPi, or Github?

**Comment by asottile** [comment]
I was hoping it could end up on `https://bootstrap.pypa.io/get-virtualenv.py` (similar to `https://bootstrap.pypa.io/get-pip.py`)

**Comment by gaborbernat** [comment]
Let's have it ready then, and we'll try to get it up there. That being said from what we talked wouldn't it be ``https://bootstrap.pypa.io/get-virtualenv.zip``?

**Comment by asottile** [comment]
could either be `.py` or `.zip` -- though I haven't done anything with [PEP 441](https://www.python.org/dev/peps/pep-0441/) yet (and from cursory glance it doesn't support py2).  I was more thinking the same approach that `get-pip.py` uses -- it embeds the `.whl` in a string literal

**Comment by ncoghlan** [comment]
The zipapp format has been supported since Python 2.6 - the only new parts in PEP 441 were the zipapp helper library for creating them, and the `pyz` file association on Windows.

**Comment by pradyunsg** [comment]
If someone does the work for packaging virtualenv in a zipapp for installation, it might be useful to consider adopting the same approach for pip too. Though, I'm not sure how the transition would be done.

"Modernize `get-pip.py`" has been in my mental TODO list since pip 10.0, but I never got to it because of the effort-reward ratio (low reward, high effort "fun" task).

**Comment by pfmoore** [comment]
For information, it's possible to bundle `virtualenv` up using the [shiv](https://pypi.org/project/shiv/) tool. I use this for a number of applications (virtualenv, tox, pytest, pipenv, ...) and it works really well.

I've not tried it for pip, but it may well work there - either for a standalone pip, or as a get-pip equivalent...

**Comment by gaborbernat** [comment]
We now have https://github.com/pypa/get-virtualenv, we just need to follow the logic from https://github.com/pypa/get-pip and deploy the zipapp generated from here 👍
